### PR TITLE
gadget/install/install.go: obey seed mounted by initramfs

### PR DESCRIPTION
**Pull request by valentindavid**
We need to be consistant on what snap-bootstrap mounts as seed and where UC is installed. There are cases where multiple disks contain UC seed. snap-bootstrap may choose a disk based on kernel parameters.

### CODE REVIEW

File: gadget/install/install.go [hunk 1]
```diff
@@ -232,6 +232,15 @@ func createPartitions(model gadget.Model, gadgetRoot, kernelRoot, bootDevice str
 
 	// auto-detect device if no device is forced
 	// device forcing is used for (spread) testing only
+	if bootDevice == "" {
+		foundDisk, err := disks.DiskFromMountPoint("/run/mnt/ubuntu-seed", nil)
+		if err == nil {
```
Comment (mvo5) [inline gadget/install/install.go:237]: (nitpick) I would invert the order here, i.e. `if err != nil {` first as it feels more idiomatic go to me (but okay to keep of course, this is really a corner case)

File: gadget/install/install.go [hunk 2]
```diff
@@ -232,6 +232,15 @@ func createPartitions(model gadget.Model, gadgetRoot, kernelRoot, bootDevice str
...
+			bootDevice = foundDisk.KernelDeviceNode()
+		} else {
+			logger.Noticef("Warning: cannot find disk from mounted seed: %s", err)
+			bootDevice = ""
```
Comment (mvo5) [inline gadget/install/install.go:241]: Unless I miss something `bootDevice`is already `""` (check above in line 235)(?)

File: gadget/install/install.go [hunk 3]
```diff
@@ -705,6 +714,15 @@ func FactoryReset(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string,
 
 	// auto-detect device if no device is forced
 	// device forcing is used for (spread) testing only
+	if bootDevice == "" {
```
Comment (mvo5) [inline gadget/install/install.go:717]: (same comments/nitpicks as above)

**Review by mvo5 — COMMENTED** [review]
Thanks, this looks good, just some minor questions/nitpicks inline :)

File: gadget/install/install.go [hunk 4]
```diff
@@ -232,6 +232,15 @@ func createPartitions(model gadget.Model, gadgetRoot, kernelRoot, bootDevice str
 
 	// auto-detect device if no device is forced
 	// device forcing is used for (spread) testing only
```
Comment (alfonsosanchezbeato) [inline gadget/install/install.go:234]: With this change, this comment is not true anymore

File: gadget/install/install.go [hunk 5]
```diff
@@ -232,6 +232,15 @@ func createPartitions(model gadget.Model, gadgetRoot, kernelRoot, bootDevice str
...
+	if bootDevice == "" {
+		foundDisk, err := disks.DiskFromMountPoint("/run/mnt/ubuntu-seed", nil)
+		if err == nil {
+			bootDevice = foundDisk.KernelDeviceNode()
+		} else {
+			logger.Noticef("Warning: cannot find disk from mounted seed: %s", err)
+			bootDevice = ""
+		}
+	}
 	if bootDevice == "" {
```
Comment (alfonsosanchezbeato) [inline gadget/install/install.go:244]: Maybe this part can be moved inside the error case above (`foundDisk.KernelDeviceNode()` should always return something if there was no error I guess?)

File: gadget/install/install.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:717]: Same comments from my part here as well. Maybe this should be a small helper function to reuse code.

**Comment by tsunghanliu** [comment]
We provided the client a test version which is built against the branch of PR and the client told that the results are positive.

> 1.First boot and factory-reset boot sequences are not corrupting the bootable SD.
> 2.eMMC partitions are getting used for mounting UC filesystem with bootable SD card.

**Review by mvo5 — APPROVED** [review]
Super nice, thank you!

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for the changes

**Pull request merged**